server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java

```diff
@@ -45,63 +15,114 @@
 
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext context) throws IOException {
-        Map<String, Query> namedQueries = new HashMap<>();
-        if (context.parsedQuery() != null) {
-            namedQueries.putAll(context.parsedQuery().namedFilters());
-        }
-        if (context.parsedPostFilter() != null) {
-            namedQueries.putAll(context.parsedPostFilter().namedFilters());
-        }
+        Map<String, Query> namedQueries = collectNamedQueries(context);
         if (namedQueries.isEmpty()) {
             return null;
         }
+
+        Map<String, Weight> weights = prepareWeights(context, namedQueries);
+
+        return context.includeNamedQueriesScore() ? createScoringProcessor(weights) : createNonScoringProcessor(weights);
+    }
+
+    private Map<String, Query> collectNamedQueries(FetchContext context) {
+        Map<String, Query> namedQueries = new HashMap<>();
+        Optional.ofNullable(context.parsedQuery()).ifPresent(parsedQuery -> namedQueries.putAll(parsedQuery.namedFilters()));
+        Optional.ofNullable(context.parsedPostFilter()).ifPresent(parsedPostFilter -> namedQueries.putAll(parsedPostFilter.namedFilters()));
+        return namedQueries;
+    }
+
+    private Map<String, Weight> prepareWeights(FetchContext context, Map<String, Query> namedQueries) throws IOException {
         Map<String, Weight> weights = new HashMap<>();
         for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
-            weights.put(
-                entry.getKey(),
-                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE_NO_SCORES, 1)
-            );
+            ScoreMode scoreMode = context.includeNamedQueriesScore() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+            weights.put(entry.getKey(), context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), scoreMode, 1));
         }
-        return new FetchSubPhaseProcessor() {
+        return weights;
+    }
 
-            final Map<String, Bits> matchingIterators = new HashMap<>();
+    private FetchSubPhaseProcessor createScoringProcessor(Map<String, Weight> weights) {
+        return new FetchSubPhaseProcessor() {
+            final Map<String, Scorer> matchingScorers = new HashMap<>();
 
             @Override
             public void setNextReader(LeafReaderContext readerContext) throws IOException {
-                matchingIterators.clear();
-                for (Map.Entry<String, Weight> entry : weights.entrySet()) {
-                    ScorerSupplier ss = entry.getValue().scorerSupplier(readerContext);
-                    if (ss != null) {
-                        Bits matchingBits = Lucene.asSequentialAccessBits(readerContext.reader().maxDoc(), ss);
-                        matchingIterators.put(entry.getKey(), matchingBits);
+                setupScorers(readerContext, weights, matchingScorers);
```

Comment:
Why we have `setupScorers` and `setupMatchingBits`? these methods are used exactly once, we could keep then in the `setNextReader`

Reply:
yeah went little bit much into the refactoring,  removing those "bit much" methods